src/lib/sw_engine/tvgSwRaster.cpp

```diff
@@ -108,6 +130,124 @@ static bool _rasterTranslucentRle(SwSurface* surface, SwRleData* rle, uint32_t c
+            auto rY = static_cast<uint32_t>(roundf((span->x + x) * invTransform.e21 + span->y * invTransform.e22 + invTransform.e23));
+
+            if (rX < 0 || rX >= width || rY < 0 || rY >= height) continue;
+
+            auto dst = &surface->buffer[span->y * surface->stride + span->x + x];
+            auto index = rY * width + rX;        //TODO: need to use image's stride
+            if (dst && data && data[index]) {
+                auto src = ALPHA_BLEND(data[index], span->coverage);
+                auto invAlpha = 255 - surface->comp.alpha(src);
+                *dst = src + ALPHA_BLEND(*dst, invAlpha);
+            }
+        }
+        ++span;
+    }
+    return true;
+}
+
+
+static bool _rasterTranslucentImage(SwSurface* surface, uint32_t *data, uint32_t opacity,const SwBBox& bbox, const Matrix* transform, uint32_t width, uint32_t height)
+{
+    Matrix invTransform;
+    _inverseMatrix(transform, &invTransform);
+    for (auto y = bbox.min.y; y < bbox.max.y; y++) {
+        for (auto x = bbox.min.x; x < bbox.max.x; x++) {
+            auto rX = static_cast<uint32_t>(roundf(x * invTransform.e11 + y * invTransform.e12 + invTransform.e13));
+            auto rY = static_cast<uint32_t>(roundf(x * invTransform.e21 + y * invTransform.e22 + invTransform.e23));
+
+            if (rX < 0 || rX >= width || rY < 0 || rY >= height) continue;
+
+            auto dst = &surface->buffer[y * surface->stride + x];
+            auto index = rX + (rY * width); //TODO: need to use image's stride
+            if (dst && data && data[index]) {
+                auto src = ALPHA_BLEND(data[index], opacity);
+                auto invAlpha = 255 - surface->comp.alpha(src);
+                *dst = src + ALPHA_BLEND(*dst, invAlpha);
+            }
+        }
+    }
+    return true;
+}
+
+
+static bool _rasterImage(SwSurface* surface, uint32_t *data, const SwBBox& bbox, const Matrix* transform, uint32_t width, uint32_t height)
+{
+    Matrix invTransform;
+    _inverseMatrix(transform, &invTransform);
+    for (auto y = bbox.min.y; y < bbox.max.y; y++) {
+        for (auto x = bbox.min.x; x < bbox.max.x; x++) {
+            auto rX = static_cast<uint32_t>(roundf(x * invTransform.e11 + y * invTransform.e12 + invTransform.e13));
+            auto rY = static_cast<uint32_t>(roundf(x * invTransform.e21 + y * invTransform.e22 + invTransform.e23));
+
+            if (rX < 0 || rX >= width || rY < 0 || rY >= height) continue;
+
+            auto dst = &surface->buffer[y * surface->stride + x];
+            auto index = rX + (rY * width); //TODO: need to use image's stride
+            if (dst && data && data[index]) {
+                auto src = data[index];
+                auto invAlpha = 255 - surface->comp.alpha(src);
+                *dst = src + ALPHA_BLEND(*dst, invAlpha);
+            }
+        }
+    }
+    return true;
+}
+
+
+bool rasterImage(SwSurface* surface, SwImage* image, uint8_t opacity, const Matrix* transform)
```

Comment:
OPTIMIZATION: if there is no any transform (identical matrix) we can draw image in the fastest track.

Reply:
changed